### PR TITLE
modification to IntervalMetadata

### PR DESCRIPTION
**Pull request by RNAer**
1. add `IntervalMetadata.num_interval_features`

2. add `upper_bound`

3. add copy methods

@jairideout , ready for review.

### CODE REVIEW

**Comment by RNAer** [comment]
@jairideout , a ping in case this escaped your attention : )

**Comment by jairideout** [comment]
Thank you for the ping. It is on my radar, but I am currently swamped with critical QIIME 2 work before "workshop season" begins. I will try my best to review this today, apologies for the delay.

File: skbio/metadata/_interval.py [hunk 1]
```diff
@@ -90,6 +91,10 @@ def __init__(self, interval_metadata, locations,
 
     def _add(self):
         """Add the current ``Interval`` to the IntervalMetadata object."""
+        if self.locations[-1][-1] > self._interval_metadata.upper_bound:
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: Should we have an explicit `lower_bound` that is always zero, then validate that here? It seems that Sequence objects require their interval metadata to be bounded by `[0, upper_bound)`, where `upper_bound` is `len(seq)`.

File: skbio/metadata/_interval.py [hunk 2]
```diff
@@ -90,6 +91,10 @@ def __init__(self, interval_metadata, locations,
...
+            raise ValueError('Cannot set `locations` (%r) with coordinate '
+                             'larger than upper bound.' % self.locations)
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: Can you include the upper bound in the error message?

File: skbio/metadata/_interval.py [hunk 3]
```diff
@@ -456,7 +467,9 @@ class IntervalMetadata():
     Interval(interval_metadata=..., locations=[(3, 9)], \
 boundaries=[(True, True)], metadata={'gene': 'sagB'})
     """
-    def __init__(self):
+    def __init__(self, upper_bound):
+        self.upper_bound = upper_bound
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: Can you make this a private attribute `self._upper_bound`, and expose it as a public, read-only property? That way users can't change it after construction (e.g. to a smaller upper bound).

File: skbio/metadata/_interval.py [hunk 4]
```diff
@@ -480,27 +499,22 @@ def inner(self, *args, **kwargs):
...
-            intvls = [(length-x[1], length-x[0]) for x in
-                      reversed(f.locations)]
+            intvls = [(self.upper_bound - x[1], self.upper_bound - x[0])
+                      for x in reversed(f.locations)]
             f.locations = intvls
 
         # DONT' forget this!!!
         self._is_stale_tree = True
 
+    @experimental(as_of='0.5.0-dev')
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: 👍 , thanks for catching these!

File: skbio/metadata/_interval.py [hunk 5]
```diff
@@ -726,6 +741,44 @@ def __repr__(self):
...
+        '''Return a deep copy.
+
+        See Also
+        --------
+        copy
+        '''
+        return self._copy(True, memo)
+
+    @experimental(as_of='0.5.0-dev')
+    def _copy(self, deep, memo):
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: Copying here is a little tricky. If we copy the `Interval` objects in this `IntervalMetadata`, they will still refer to this object as their parent. Any subsequent updates to these copied `Interval` objects will only affect *this* `IntervalMetadata` object, while we'd want those updates to be reflected in both the current object and the copied object (e.g. staled tree state). I don't think this behavior is possible with the current design because each `Interval` has only a single parent.

Since `IntervalMetadata` and `Interval` are inherently linked (bidirectional), we need to deep copy both types of objects. The only part of `Interval` that isn't necessarily deep-copied is the `.metadata` dict, since that can be safely shallow- or deep-copied using `copy.copy` or `copy.deepcopy`, respectively.

Here's one way to do this:

```python
cp = IntervalMetadata(self.upper_bound)

for interval in self._intervals:
    # Only need to shallow-copy `locations` and `boundaries` because their elements are immutable.
    locations_cp = interval.locations[:]
    boundaries_cp = interval.boundaries[:]
    if deep:
        metadata_cp = copy.deepcopy(interval.metadata, memo)
    else:
        metadata_cp = copy.copy(interval.metadata)
    
    cp.add(locations_cp, boundaries=boundaries_cp, metadata=metadata_cp)

return cp
```

Does this copying strategy make sense?

Once we've finalized the copying behavior, can you also update the unit tests so that they would have failed with the current implementation?

File: skbio/metadata/_interval.py [hunk 6]
```diff
@@ -726,6 +741,44 @@ def __repr__(self):
...
+    def __deepcopy__(self, memo):
+        '''Return a deep copy.
+
+        See Also
+        --------
+        copy
+        '''
+        return self._copy(True, memo)
+
+    @experimental(as_of='0.5.0-dev')
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: Remove API stability tag, it's only applied to public API

File: skbio/metadata/_interval.py [hunk 7]
```diff
@@ -90,6 +91,10 @@ def __init__(self, interval_metadata, locations,
 
     def _add(self):
         """Add the current ``Interval`` to the IntervalMetadata object."""
+        if self.locations[-1][-1] > self._interval_metadata.upper_bound:
+            raise ValueError('Cannot set `locations` (%r) with coordinate '
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: Can you add unit tests for adding `Interval` objects that fall outside the `IntervalMetadata` range?

**Comment by jairideout** [comment]
@RNAer I'm done reviewing, thanks for working on this!

File: skbio/metadata/_interval.py (hunk 1 above)
Comment (RNAer) [inline skbio/metadata/_interval.py]: I added to check the lower bound in line 250. I also moved the check into `_locations_boundaries_setter` so the `locations` setter does not escape the check.

You were not talking about adding `lower_bound` as an attribute, right? Because in all of our use cases, it should always start with 0; otherwise, it needs other API change.

**Comment by RNAer** [comment]
@jairideout , thanks for reviewing! I think I addressed you comments. Nice catch on the `copy` issue.

File: skbio/metadata/_interval.py (hunk 1 above)
Comment (jairideout) [inline skbio/metadata/_interval.py]: It would be nice to have a read-only `lower_bound` property that always returns zero, just for symmetry with the `upper_bound` property. Users wouldn't be able to configure this but it could be possible in the future.

**Comment by jairideout** [comment]
Thanks @RNAer for addressing the comments! I just replied to your question about a `lower_bound` property. Let me know when that's done and I'll take another pass tomorrow.

File: skbio/metadata/_interval.py (hunk 1 above)
Comment (RNAer) [inline skbio/metadata/_interval.py]: not adding `lower_bound` as a param to the `IntervalMetadata.__init__`, right? I really hesitate to expose it in the constructor for users to set it.
```
def __init__(self, upper_bound):
     self._lower_bound = 0
     ...
```

Comment (jairideout) [inline skbio/metadata/_interval.py]: Right, this would just be a read-only property that users can't configure

Comment (jairideout) [inline skbio/metadata/_interval.py]: It doesn't even need to be an instance attribute, this is sufficient:

```python
@property
def lower_bound(self):
    return 0
```

**Comment by RNAer** [comment]
@jairideout , ready for another round. thanks for your comments!

**Comment by RNAer** [comment]
good point! changed accordingly, @jairideout

**Comment by RNAer** [comment]
@jairideout , we know you guys are swamped with QIIME2 (which looks awesome, BTW); is there anyway we can help from our end speed up the code review process?

**Comment by jairideout** [comment]
@RNAer, I was planning to review it today and will get to it before the end of the day. I have not forgotten about this PR.

@gregcaporaso has given me an hour a day, or a single review passthrough (whichever is less time) to spend on interval metadata PRs. He has also stated that it is okay if I can't get to the PR each day due to other competing tasks. I am trying my best to be responsive and keep the review process moving forward, and have usually been following up on this PR and #1423 within a day after you have posted comments or pushed changes. I have even reviewed these PRs outside of my work week/hours to keep it moving forward. For other open source projects, this would be considered a "responsive" review process (especially considering I am the only one reviewing it).

There isn't anything we can do to speed up the review process until November 7th, when we will have finished teaching the first two QIIME 2 workshops. The tasks I'm working on to get QIIME 2 to a teachable state takes highest priority until November 7th. If the interval metadata functionality is critical for your application, please implement it there (we've discussed options for this in the past) so that you aren't blocked.

**Comment by RNAer** [comment]
@jairideout , you are doing a great job and responsive considering all the other things happening at the same time. That's why I asked what we can do from our end, not your end, to speed it up. 

The problem with doing it on a divergent branch is that it complicates the future merge. It makes all the code built upon this susceptible to changes upon this future merge. GIven the amount of unit tests and documentation associated with code, I view this is a waste of efforts to build things upon something may change in soon future.

If you think otherwise, pls don't hesitate to let me know. : )

**Comment by jairideout** [comment]
> @jairideout , you are doing a great job and responsive considering all the other things happening at the same time.

Thanks @RNAer, I just wanted to be clear about my time commitment to this over the next several weeks.

> That's why I asked what we can do from our end, not your end, to speed it up.

Having more reviewers on your end will help speed this up. I'm about to take a pass through now and I think this one is close to being ready, so probably not worth having more reviewers on this PR.

> The problem with doing it on a divergent branch is that it complicates the future merge. It makes all the code built upon this susceptible to changes upon this future merge. GIven the amount of unit tests and documentation associated with code, I view this is a waste of efforts to build things upon something may change in soon future.
>
>If you think otherwise, pls don't hesitate to let me know. : )

I agree it would be ideal to have the code your application needs available in scikit-bio, in order to avoid writing code that your application needs presently *and* also porting it to scikit-bio. However, we are currently limited in code reviewer bandwidth and core maintainers of this project, and regardless of that, it takes a lot of time to get pull requests to a state where they can be merged into the package. As you noted, code contributions are accompanied by extensive unit tests and documentation, and that adds a lot of time to the review.

While scikit-bio is a newer player in the SciPy stack, this situation is analogous to needing a critical feature for your application *now*, and relying on that feature being merged into a codebase such as numpy, scipy, pandas, etc. before continuing forward with your application development. As we saw with pandas IntervalIndex, some code never even makes it into a package, or is delayed for some time. An extreme example of this is the addition of `os.scandir` to the Python standard library: [it took years](http://benhoyt.com/writings/scandir/) to get the function added. Now, we obviously move faster than that, but the developer of `os.scandir` created his own package while waiting for his function to be merged into Python. This strategy can be applied to any type of project.

We are excited to have interval metadata support in scikit-bio, but things will move more slowly than implementing this functionality in your application or as a separate package. Ideally, all useful bioinformatics functionality will be accessible through scikit-bio someday, but it'll take time to get there.

File: skbio/metadata/_interval.py [hunk 8]
```diff
@@ -726,6 +765,50 @@ def __repr__(self):
 
         return '\n'.join([l1, l2] + items)
 
+    @experimental(as_of='0.5.0-dev')
+    def __copy__(self):
+        '''Return a shallow copy.
+
+        Notes
+        -----
+        The references to original ``Interval`` objects are created
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: I don't think this is true anymore. Now deep and shallow copy will always create a new `IntervalMetadata` object with new `Interval` objects. The metadata dict is the only thing that is shallow vs. deep copied.

Can you update this docstring, as well as the deepcopy docstring, to describe this behavior? Something along the lines of:

(For shallow copy)

"The `IntervalMetadata` copy will have copies of the `Interval` objects present in this `IntervalMetadata` object. Each `Interval` object's `.metadata` dictionary will be a shallow copy."

(For deep copy)

"The `IntervalMetadata` copy will have copies of the `Interval` objects present in this `IntervalMetadata` object. Each `Interval` object's `.metadata` dictionary will be a deep copy."

File: skbio/metadata/_interval.py [hunk 9]
```diff
@@ -726,6 +765,50 @@ def __repr__(self):
...
+        in the returned copy object.
+
+        See Also
+        --------
+        deepcopy
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: change to `__deepcopy__`

File: skbio/metadata/_interval.py [hunk 10]
```diff
@@ -726,6 +765,50 @@ def __repr__(self):
...
+        '''
+        return self._copy(False, {})
+
+    @experimental(as_of='0.5.0-dev')
+    def __deepcopy__(self, memo):
+        '''Return a deep copy.
+
+        See Also
+        --------
+        copy
```
Comment (jairideout) [inline skbio/metadata/_interval.py]: change to `__copy__`

**Comment by RNAer** [comment]
@jairideout , corrected the docstring.

Thanks for sharing that `os.scandir` blog - it is a fun read!

Maybe this is not the right place to ask this, but you and @ebolyen are the only core maintainers currently? What qualifies additional people to be in the core? As scikit-bio is gaining momentum, it seems more hands would be really helpful?

**Comment by jairideout** [comment]
# tl;dr

We haven't formalized who the core maintainers are, or how to become a core maintainer. This is something we are planning to formalize sometime in November or December after the QIIME 2 deadlines. We absolutely want, and need, more core maintainers.

## My thoughts
Here are my thoughts (also discussed with @ebolyen and @thermokarst):

Currently, I would think of @gregcaporaso, @ebolyen, and I as core maintainers due to our contributions to the overall structure, design, and direction of the project, particularly in the past two years. For example, the I/O registry, sequence subpackage rewrite, redesign of public APIs to be consistent across the codebase, API stability, private/package-private/public API distinction, documentation system and standards, release management, code reviews, and public outreach. In essence, we've redesigned the package such that it no longer resembles a port of PyCogent and has an entirely new direction/philosophy in place.

I think a core maintainer:

- is someone who has contributed to the project enough to have a comprehensive understanding of the project's structure and API design across the codebase, and how new code/doc/test contributions should best fit into this structure.

- understands the direction of the project enough to determine whether a contribution is appropriate to include or not.

- understands the level of standards set in place for the project, and how to get contributions to meet those standards.

- can determine when it is appropriate to request the contributor to write an RFC, and is able to guide an RFC through its lifecycle to determine whether it should be accepted or rejected.

A core maintainer doesn't need to have specific knowledge of the internals of each module/function/class/etc, but needs to be able to identify the appropriate maintainer of that functionality when questions arise or the core maintainer needs other input. For example, I am not familiar enough with the math in the `skbio.stats.composition` module @mortonjt wrote to field questions like #1393 or #1415, but I am aware that @mortonjt is the maintainer and am able to direct specific questions to him.

With code reviews, it is helpful to have more reviewers (core maintainers or not) taking passes through the code. Anyone is welcome to review. This will help a lot with reducing the time it takes to get something merged. However, at least one core maintainer needs to review the code to ensure the higher-level concerns mentioned above are addressed, and a core maintainer should sign off on the PR and merge.

I realize many of the things I've described are pretty subjective. We will do our best to formalize what makes a core maintainer and how to become one. I am also planning on writing a code review checklist doc to help other code reviewers perform in-depth reviews necessary for this package, so that I am not the bottleneck in getting things merged. I am also planning to improve checklist.py to verify some of the things I have to manually verify during review, which will help other reviewers too.

These are just my thoughts on the subject of core maintainers; nothing has been finalized and this all warrants further discussion as a group. It is not our intent to be exclusive here, and once formalized, we plan to train interested developers to expand the pool of core maintainers and reviewers.

**Comment by antgonza** [comment]
I guess this text can be used as the beginning of an specific issue to discuss this. Also, from that text you already defined 3 roles as far as I can tell (which is cool): core dev, dev, and reviewer.
